source/code/plugin/customtailfilereader.rb

```diff
@@ -0,0 +1,389 @@
+require 'optparse'
+require 'json'
+require 'logger'
+require_relative 'omslog'
+require 'fluent/filter'
+
+module CustomTailscript
+
+  class CustomNewTail  
+    def initialize(paths)
+      @paths = paths
+      @tails = {}
+      @pos_file = $options[:pos_file] 
+      @read_from_head = $options[:read_from_head]
+      @pf = nil
+      @pf_file = nil
+
+      @log = Logger.new(STDERR)
+      @log.formatter = proc do |severity, time, progname, msg|
+        "#{severity} #{msg}\n"
+      end
+    end
+
+    attr_reader :paths
+
+    def start
+      start_watchers(@paths) unless @paths.empty?
+    end
+
+    def shutdown
+      @pf_file.close if @pf_file
+    end
+
+    def setup_watcher(path, pe)
+      tw = TailWatcher.new(path, pe, @read_from_head, @log, &method(:receive_lines))
+      tw.on_notify
+      tw
+    end
+
+    def start_watchers(paths)
+      if @pos_file
+        @pf_file = File.open(@pos_file, File::RDWR|File::CREAT)
+        @pf_file.sync = true
+        @pf = PositionFile.parse(@pf_file)
+      end
+      paths.each { |path|
+        pe = nil
+        if @pf
+          pe = @pf[path]    #pe is FilePositionEntry instance
+          if pe.read_inode.zero?
+            begin
+              pe.update(File::Stat.new(path).ino, 0)
+            rescue Errno::ENOENT
+              @log.warn "#{path} not found. Continuing without tailing it."
+            end
+          end
+        end
+        
+        @tails[path] = setup_watcher(path, pe)
+      }
+    end
+
+    def receive_lines(lines, tail_watcher)
+      unless lines.empty?
+        puts lines
+      end
+      return true
+    end
+
+    class TailWatcher
+      def initialize(path, pe, read_from_head, log, &receive_lines)
+        @path = path
+        @pe = pe || MemoryPositionEntry.new
+        @read_from_head = read_from_head
+        @log = log
+        @receive_lines = receive_lines
+        @rotate_handler = RotateHandler.new(path, log, &method(:on_rotate))
+        @io_handler = nil
+      end
+
+      attr_reader :path
+
+      def wrap_receive_lines(lines)
+	newLines = []
+	containerID = @path.split('/').last.chomp('-json.log')
+	containerInspectInformation = "/var/opt/microsoft/docker-cimprov/state/ContainerInventory/" + containerID
+	tempContainerInfo = {}
+	File.open(containerInspectInformation) { |f| tempContainerInfo = JSON.parse(f.readline)}
```

Comment:
Add exception handling , for file open fail, json parse fail and catch all other exceptions

Reply:
Add exception handling for this.